Extend (and change) the definition of links endpoints (#273)
Materials-Consortia/OPTIMADE: optimade.rst

Links Endpoint
--------------

This endpoint exposes information on other OPTIMADE API implementations that are related to the current implementation.
The links endpoint MUST be provided under the versioned base URL at :endpoint:`/links`.

Link Types
~~~~~~~~~~

Each link has a :property:`link_type` attribute that specifies the type of the linked relation.

The :property:`link_type` MUST be one of the following values:

- :field-val:`child`: a link to another OPTIMADE implementation that MUST be within the same provider.
  This allows the creation of a tree-like structure of databases by pointing to children sub-databases.
- :field-val:`root`: a link to the root implementation within the same provider.
  This is RECOMMENDED to be an `Index Meta-Database`_.

>   There MUST be only one :val:`root` implementation per provider and any :val:`child` (or :val:`child` of a :val:`child`, at any depth) MUST have a link to the :val:`root` implementation.

Reviewer: Do I read this correctly that the root itself is formally exempt from having a root link? I think it is an odd quirk to not just say that the 'root' link is always mandatory (and thus the root links to itself). So:
```suggestion
  There MUST be only one :val:`root` implementation per provider and all implementations MUST have a link to this :val:`root` implementation. If the provider only supply a single implementation, the :val:`root` link links to the implementation itself.
```

Reviewer 2: I see this is an unintended meaning of this sentence. The `root` implementation can of course have a `root` links resource (inevitably pointing to itself).

>   There MUST be only one :val:`root` implementation per provider and all implementations MUST have a link to this :val:`root` implementation. If the provider only supply a single implementation, the :val:`root` link links to the implementation itself.

Reviewer 2: Splitting to one sentence per one line and fixing a minor grammatical mistake.
```suggestion
  There MUST be only one :val:`root` implementation per provider and all implementations MUST have a link to this :val:`root` implementation.
  If the provider only supplies a single implementation, the :val:`root` link links to the implementation itself.
```

        },
        {
          "type": "links",
          "id": "frameworks",
          "attributes": {
            "name": "Some other DB",
            "description": "A DB by the example2 provider",
            "base_url": "http://example2.com/some_db/optimade",
            "homepage": "http://example2.com",
            "link_type: "external"
          }
        },
        {
          "type": "links",
          "id": "optimade",
          "attributes": {
            "name": "Materials Consortia",
            "description": "List of OPTIMADE providers maintained by the Materials Consortia organisation",
            "base_url": "https://providers.optimade.org",
            "homepage": "https://optimade.org",
            "link_type": "providers"
          }
        }
      // ... <other objects>
      ]
      // ...
    }

Internal Links: Root and Child Links
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Any number of resource objects with :property:`link_type`=:val:`child` MAY be present as part of the :field:`data` list.
A :val:`child` object represents a "link" to an OPTIMADE implementation within the same provider exactly one layer **below** the current implementation's layer.

> Either none or a single resource object with :property:`link_type`=:val:`root` MAY be present as part of the :field:`data` list.

Reviewer: ```suggestion
Exactly one resource object with :property:`link_type`=:val:`root` MUST be present as part of the :field:`data` list.
```
So, see above, I think exactly one root link should always be present. That is easier to validate, and a client can always rely on finding the root from any starting database by following the root link once.

Reviewer 2: It's a bit more stringent, but fine by my accounts. I would like more thumbs up than just my own on this though.

Author: Ok with me!

Note: the same implementation may of course be linked by other implementations via a ' :endpoint:`/links` endpoint with :property:`link_type`=:val:`external`.

The :val:`root` resource object represents a link to the topmost OPTIMADE implementation of the current provider.
By following :val:`child` links from the :val:`root` object recursively, it MUST be possible to reach the current OPTIMADE implementation.

In practice, this forms a tree structure for the OPTIMADE implementations of a provider. 
**Note**: The RECOMMENDED number of layers is two.

List of Providers Links
~~~~~~~~~~~~~~~~~~~~~~~

> Resource objects with :property:`link_type`=:val:`providers` links to an `Index Meta-Database`_ that supplies a list of OPTIMADE database providers.

Reviewer 2: MUSTy addition
```suggestion
Resource objects with :property:`link_type`=:val:`providers` links MUST point to an `Index Meta-Database`_ that supplies a list of OPTIMADE database providers.
```

Author: Ok with me

> Resource objects with :property:`link_type`=:val:`providers` are meant to indicate links to an `Index Meta-Database`_ hosted by organisations that want to point to various OPTIMADE database providers.

Reviewer: Am I not allowed to link to an index meta-database of providers unless it is maintained by an organization?:
```suggestion
Resource objects with :property:`link_type`=:val:`providers` links to an `Index Meta-Database`_ that supplies a list of OPTIMADE database providers.
```
Also, should there be a MUST somewhere here?
